internal/util/pool.go

```diff
@@ -6,105 +6,171 @@ import (
-// A cancellation will return (nil, nil, nil).
-// Consult the returned map of errors using the item's index to check for item-level errors.
+// every input will be in PoolResult.Results.
+// A cancellation will stop new work from starting but complete in progress work asynchronously.
+// The partial results from a cancellation are available after PoolResult.IsDone() returns.
+// Consult the returned map of errors using the item to check for item-level errors.
+// Work level errors will not generate a critical error, so both must be checked independently.
 // The underlying collection of input work items must not be modified as this function is running.
-func RunParallelPoolMap[InputType any, OutputType any](poolSize int, workItems []InputType, ctx context.Context, workFunc func(InputType) (OutputType, error)) ([]OutputType, map[int]error, error) {
+func RunParallelPoolMap[InputType comparable, OutputType any](poolSize int, workItems []InputType, ctx context.Context, workFunc func(InputType) (OutputType, error)) (PoolResult[InputType, OutputType], error) {
 	if poolSize <= 0 {
-		return nil, nil, fmt.Errorf("Pool size must be positive, got %d.", poolSize)
-	}
-
-	type WorkItem struct {
-		Index int
-		Item  InputType
+		return PoolResult[InputType, OutputType]{}, fmt.Errorf("Pool size must be positive, got %d.", poolSize)
 	}
 
 	type ResultItem struct {
-		Index int
-		Item  OutputType
-		Error error
+		Input  InputType
+		Result OutputType
+		Error  error
 	}
 
-	results := make([]OutputType, len(workItems))
-	workErrors := make(map[int]error)
-
 	resultQueue := make(chan ResultItem, poolSize)
-	workQueue := make(chan WorkItem, poolSize)
-	doneChan := make(chan bool, poolSize)
-	exitWaitGroup := sync.WaitGroup{}
+	workQueue := make(chan InputType, poolSize)
+	doneChan := make(chan any)
+	doneWorkingChan := make(chan any)
+
+	workerExitWaitGroup := sync.WaitGroup{}
+	threadExitWaitGroup := sync.WaitGroup{}
+
+	workerExitWaitGroup.Add(poolSize)
+	// There will always be poolSize + 2 threads started.
+	// A thread will be started for the work loader, the work collecter, and each worker in the pool.
+	// Every one of these work threads must call threadExitWaitGroup.Done() to avoid a deadlock.
+	threadExitWaitGroup.Add(poolSize + 2)
+
+	output := PoolResult[InputType, OutputType]{
+		Results:    make(map[InputType]OutputType, len(workItems)),
+		WorkErrors: make(map[InputType]error, 0),
+		done:       doneChan,
+	}
 
 	// Load work.
 	go func() {
-		for i, item := range workItems {
+		defer threadExitWaitGroup.Done()
+
+		for _, item := range workItems {
 			// Either send a work item, or cancel.
 			select {
 			case <-ctx.Done():
 				return
-			case workQueue <- WorkItem{i, item}:
+			case workQueue <- item:
 				// Item was already sent on the chan, do nothing here.
 			}
 		}
+
+		// Signal to the workers that there is no more work.
+		close(workQueue)
```

Comment:
This is not hit in the cancellation case.
What about putting it in a defer with `threadExitWaitGroup.Done()`?